File: sorting & searching/C++/merge_sort.cc
/**
 * @file merge_sort.cc
 * @author Hongyu Ding
 * @brief This is the code of merge sort
 * @version 1.0
 * @date 2020-10-23
 * 
 * @copyright Copyright (c) 2020
 * 
 */

#include <iostream>
#include <vector>

void merge(std::vector<int> &a, std::vector<int> &l, std::vector<int> &r) {
    int i = 0, j = 0, index = 0;
    while (i < l.size() || j < r.size()) {
        if (i < l.size() && j < r.size()) {
            if (l[i] < r[j]) {
                a[index] = l[i]; i ++;
            } else {
                a[index] = r[j]; j ++;
            }
        } else if (i == l.size()) {
            while (j < r.size()) {
                a[index] = r[j];
                j ++; index ++;
            }
            break;
        } else if (j == r.size()) {
            while (i < l.size()) {
                a[index] = l[i];
                i ++; index ++;
            }
            break;
        }
        index ++;
    }
}

/**
 * @brief merge_sort applies merge_sort on array a
 * O(n log n)
 * 
 * @param a array
 */
void merge_sort(std::vector<int> &a) {
    if (a.size() <= 1) return;

    std::vector<int> l, r;
    int mid = (static_cast<int>(a.size()) + 1) / 2;

    for (int i = 0; i < mid; i ++) l.emplace_back(a[i]);
    for (int i = mid; i < static_cast<int>(a.size()); i ++) r.emplace_back(a[i]);

    merge_sort(l);
    merge_sort(r);
    merge(a, l, r);
}

int main() {
    std::vector<int> array;
    int n;
    
    // input
    std::cin >> n;
    for (int i = 0; i < n; i++) {
        int val;
        std::cin >> val;
        array.emplace_back(val);
    }

    merge_sort(array);

    // output
    for (int i = 0; i < n; i++) {
        std::cout << array[i] << " ";
    }
    std::cout << std::endl;

    return 0;
}

File: sorting & searching/Java/MergeSort.java
/**
 *
 * @File Merge Sort
 * @Author Weilan Tao
 * @Date 2020-10-23
 * @Description This is the code for merge sort, java version.
 * @Since version-1.1
 * @Copyright Copyright (c) 2020
 */


import java.util.Arrays;
import java.util.Scanner;

public class MergeSort {

    /**
     * merge sort
     *
     * @param array
     * @param len
     * O(nlogn) in worst case
     */
    public static void mergeSort(int[] array, int len) {
        if (len > 1) {
            int mid = len / 2;
            mergeSort(array, mid, len);
        } else ;
    }

    /**
     * recursion part of the merge sort， return merged sub-array
     *
     * @param array
     * @param mid
     * @param len
     */
    public static void mergeSort(int[] array, int mid, int len) {
        if (len == 1) ;
        else {
            int[] l = new int[mid];
            int[] r = new int[len - mid];
            int lmid = mid / 2;
            int rmid = (len - mid) / 2;
            System.arraycopy(array, 0, l, 0, mid);
            System.arraycopy(array, mid, r, 0, len - mid);

            mergeSort(l, lmid, mid);
            mergeSort(r, rmid, len - mid);
            merge(l, r, array);
        }
    }

    /**
     * merge two sorted sub-arrays
     *
     * @param l the left sorted array
     * @param r the right sorted array
     * @param a the target merged array
     */
    public static void merge(int[] l, int[] r, int[] a) {
        int i = 0;
        int j = 0;
        int k = 0;

        while (i < l.length && j < r.length) {
            if (l[i] < r[j]) {
                a[k] = l[i];
                i++;
            } else {
                a[k] = r[j];
                j++;
            }
            k++;
        }

        while (i < l.length) {
            a[k] = l[i];
            i++;
            k++;
        }

        while (j < r.length) {
            a[k] = r[j];
            j++;
            k++;
        }

    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        System.out.println("enter the length of an array");
        int x = scanner.nextInt();

        int[] array = new int[x];
        System.out.println("set the array");

        for (int i = 0; i < x; i++) {
            array[i] = scanner.nextInt();
        }

        mergeSort(array, x);

        System.out.println("print the sorted array" + Arrays.toString(array));
        scanner.close();
    }
}
